### ISSUE TITLE
[doc] is there any performance tips on using cutorch and cunn?

### ISSUE THREAD

**Issue by cdluminate** [torch/cutorch#497]
Hi torch developers,

I'm trying to train a hierarchical (and dynamic) model on GPU but the preliminary result is disappointing --
the CPU version forward+backward costs 10 seconds but the CUDA version consumes more than 3 minutes for forward+backward. I inspected the GPU ram usage with `watch -n0.05 nvdia-smi` and the memory usage curve is drastically oscillating ... I guess the huge amount of `cudaMemcpy` and `CudaFree` calls are the major reason of low performance, but I need to look into it deeper.

cutorch and cunn are easy to use, and user and ship their model/data onto GPU with simply a `:cuda()` call,
however is redundant `:cuda()` call harmful? e.g.

``` lua
local data = torch.Tensor(SORT_OF_DATA) -- FloatTensor
data = data:cuda() -- ship data onto GPU
data = data:cuda() -- is it harmful to do it many times? I mean will it "reshuffle" tensor in the GPU memory?
```

Apart from that, I'd like to ask if there are some performance tips and good practice for the torch community on how to make good use of cuda instead of struggling with GPU memory control?

**Comment by fmassa** [comment]
Calling `:cuda()` a second time doesn't allocate new memory, so this part is ok.
I'd advise you to avoid allocating memory at each forward/backward call. This is the current pattern that is followed by `nn`/`cunn`. So, instead of doing something like
```lua
local res = a:cuda() * b:cuda() + 1
```
which performs several allocations (converting a to gpu, converting `b` to gpu, allocating `a*b`, allocating the previous result and adding 1). Instead, you could do something like
```lua
self.res = self.res or torch.CudaTensor()
self.aCuda = self.aCuda or torch.CudaTensor()
self.bCuda = self.bCuda or torch.CudaTensor()
self.aCuda:resize(a:size()):copy(a)
self.bCuda:resize(b:size()):copy(b)
self.res:cmul(self.aCuda, self.bCuda):add(1)
```
This will already give some great speedups.
Then, another thing that might be hurting performance is if you are accessing individual GPU elements and modifying them. Avoid doing something like
```lua
for i = 1, a:numel() do
  a[i] = a[i]*a[i] + 2 -- this is super slow on the GPU
end
```

**Comment by CDLuminate** [comment]
@fmassa Thank you for the interesting and inspiring answer. I then wrote some stuff
```lua
#!/usr/bin/th
-- cuda performance tester
-- https://github.com/torch/cutorch/issues/497

require 'torch'

local huge = 2500

print 'CPU mode reference'
local weight = torch.randn(256, 4096)
local timer = torch.Timer()
for iter = 1, huge do
   local data = torch.rand(4096, 8) -- batchsize 8
   local res = weight * data
end
print(timer:stop():time())

require 'cutorch'
print 'CUDA mode 1 -- cudamemcpy every iteration by mistake'

weight = weight:float() -- deliberate, emulating negligence
timer = timer:new()
for iter = 1, huge do
   local data = torch.rand(4096, 8) -- batchsize 8
   local res = torch.mm(weight:cuda(), data:cuda())
end
print(timer:stop():time())

print 'CUDA mode 2 -- normal'
weight = weight:cuda()
timer = timer:new()
for iter = 1, huge do
   local data = torch.rand(4096, 8):cuda() -- batch 8
   local res = torch.mm(weight, data)
end
print(timer:stop():time())

print 'CUDA mode 3 -- pre-create cudatensor, reducing sys time'
timer = timer:new()
weight = weight:cuda()
local mydata = torch.CudaTensor()
local myweight = torch.CudaTensor()
local myres = torch.CudaTensor()
for iter = 1, huge do
   local data = torch.rand(4096, 8)
   myweight:resize(weight:size()):copy(weight)
   mydata:resize(data:size()):copy(data)
   myres = torch.mm(myweight, mydata)
end
print(timer:stop():time())
```
and the result is also interesting
```
CPU mode reference	
{
  real : 5.7603139877319
  sys : 0.3
  user : 5.459999
}
CUDA mode 1 -- cudamemcpy every iteration by mistake	
{
  real : 7.390741109848
  sys : 0.783333
  user : 6.606666
}
CUDA mode 2 -- normal	
{
  real : 1.5737948417664
  sys : 0.133334
  user : 1.44
}
CUDA mode 3 -- pre-create cudatensor, reducing sys time	
{
  real : 1.4949600696564
  sys : 0.01
  user : 1.483333
}
```
Compared the sys time of cuda mode 2 and cuda mode 3, I noticed the the sys time is reduced. Is my understanding correct?

**Comment by fmassa** [comment]
@CDLuminate One more thing that you need to have in mind.

Most of `cutorch` operations are asynchronous, meaning that once you run an operation it immediately returns.
Your benchmarkings might be off because of that, and you should probably add a `cutorch.synchonize()` before measuring the times.

Also, note that you are still allocating memory at every iteration in mode 3. Indeed, the return value `myres` is being allocated over and over.

**Comment by CDLuminate** [comment]
This example is easy to understand:
```lua
for i = 1, a:numel() do
  a[i] = a[i]*a[i] + 2 -- this is super slow on the GPU
end
```

while with cuda kernel
```cuda
__global__ void
myvectorOperation(const float *A, int numElements)
{
    int i = blockDim.x * blockIdx.x + threadIdx.x;
    if (i < numElements) {
        A[i] = A[i] * A[i] + 2.;
    }
}
```
the operation complexity reduced from `O(n)` to `O(1)`.
so it is suggested to write it like this
```lua
a:pow(2):add(1)
```

**Comment by CDLuminate** [comment]
@fmassa Updated the script, is this better?
```lua
#!/usr/bin/th
-- cuda performance tester
-- https://github.com/torch/cutorch/issues/497

require 'torch'
require 'cutorch'

local huge = 2500
local weight = torch.randn(256, 4096):cuda()

print 'CUDA mode 2 -- normal'

cutorch.synchronize();
local timer = torch.Timer()

for iter = 1, huge do
   local data = torch.rand(4096, 8):cuda() -- batch 8
   local res = torch.mm(weight, data)
end

cutorch.synchronize();
print(timer:stop():time())

print 'CUDA mode 3 -- pre-create cudatensor, reducing sys time'

local mydata = torch.CudaTensor():resize(4096, 8)
local myweight = torch.CudaTensor():resize(weight:size())
local myres = torch.CudaTensor():resize(256, 8)

cutorch.synchronize();
timer = timer:new()

for iter = 1, huge do
   local data = torch.rand(4096, 8) -- batch 8
   myweight:copy(weight)
   mydata:copy(data)
   torch.mm( myres, myweight, mydata ) -- C <- A * B
end

cutorch.synchronize();
print(timer:stop():time())
```
with result
```
CUDA mode 2 -- normal	
{
  real : 2.1208629608154
  sys : 0.21
  user : 1.913333
}
CUDA mode 3 -- pre-create cudatensor, reducing sys time	
{
  real : 1.5044639110565
  sys : 0.046667
  user : 1.456667
}
```

**Comment by CDLuminate** [comment]
I have some more findings when trying to optimize my program:

1. use `torch.Timer` to measure time cost instead of `xlua.progress` whose time report is processed.

2. try to do the same thing with less code.

3. try to create as few `Tensor` variables as possible

and these are the conclusion from the above discussion:

1. reuse `torch.Tensor` if possible, avoiding re-creating tensors.

2. redundant `:cuda()` calls won't harm.

3. try to use `tensor:add(tensor2)` (inplace) instead of `tensor + tensor2` which creates new tensor.

4. don't apply operation of `CudaTensor` element by element, do all of it at once if possible. If really need to do that, ship the tensor back to CPU.

And seems that many of them works for CPU version too.

and this problem is also interesting:

1. when should I consider using `torch/threads`?

**Comment by CDLuminate** [comment]
I implemented a variant of Contrastive loss as follows, and optimized as suggested by fmassa . However its backward process takes `0.040` seconds on CUDA while `0.006` seconds on CPU.

I have forced shipping tensors that must be modified element-by-element on CPU, and  now I feel it hard to optimize further ... So I'm thinking of rewritting my loss function in C, because C implementation has explicit type declaration (about exactly what is on CPU, and what is on GPU, e.g. `THTensor_*`, `THCudaTensor_*`), but I'm not sure how much speed boost up I can obtain.

 @fmassa @soumith May I ask for some advise?

BTW, according to fmassa's suggestion I reduced about 3 seconds (10.5 sec -> 7.6 sec) each iteration, thanks!

```lua
local crit, parent = torch.class('nn.ContrastiveLoss', 'nn.Criterion')

[...]

function crit:updateGradInput(input, gradInput, scale)
   assert(input.cnnfeat, 'cnnfeat missing')   -- single CNN  feature
   assert(input.lstmfeat, 'lstmfeat missing') -- single LSTM feature
   assert(input.all_cnnfeats, 'all_cnnfeats missing')   -- all CNN  features of batch
   assert(input.all_lstmfeats, 'all_lstmfeats missing') -- all LSTM features of batch
   assert(input.iid, 'iid missing') -- not used
   assert(input.pn, 'pn missing')   -- not used
   local margin = self.margin

   local feats_num = #input.all_cnnfeats
   assert(#input.all_cnnfeats == #input.all_lstmfeats)

   local cnn_feats_dim = input.all_cnnfeats[1]:size(1)
   local lstm_feats_dim = input.all_lstmfeats[1]:size(1)
   assert(cnn_feats_dim == lstm_feats_dim)

   self.ins_cnnfeat:resize(1, cnn_feats_dim):copy(input.cnnfeat)
   self.ins_lstmfeat:resize(1, cnn_feats_dim):copy(input.lstmfeat)
   if self.gpuid >= 0 then
      self.ins_cnnfeat = self.ins_cnnfeat:cuda();
      self.ins_lstmfeat = self.ins_lstmfeat:cuda();
   end

   -- get myid and myskip
   local myid = 0 -- 1 <= myid <= k
   for i = 1, feats_num do
      -- find myself via one-by-one comparison, to obtain myid
      if input.lstmfeat:view(1, lstm_feats_dim):eq(
            input.all_lstmfeats[i]:view(1, lstm_feats_dim)):sum() == input.lstmfeat:nElement() then
         myid = i
         break
      end
   end
   assert(myid ~= 0, 'who am i? cannot get myid')
   local myskip = self.skipmask[{ myid, {} }]:view(1, feats_num):byte() -- force CPU
   assert(torch.type(myskip)=='torch.ByteTensor', 'low perf')

   local function Token1 (margin, x, v, vk)
      return (margin - x:dot(v) + x:dot(vk)) -- scalar
   end
   local function Token2 (margin, v, x, xk)
      return (margin - v:dot(x) + v:dot(xk)) -- scalar
   end

   self.pebpxi:resize(1, cnn_feats_dim):zero()
   self.pebpvi:resize(1, cnn_feats_dim):zero()
   if self.gpuid >= 0 then self.pebpxi = self.pebpxi:cuda(); end
   if self.gpuid >= 0 then self.pebpvi = self.pebpvi:cuda(); end
   for k = 1, feats_num do
      if myid ~= k then
         if myskip[{ 1, k }] ~= 1 then
            if Token1(margin, self.ins_cnnfeat, self.ins_lstmfeat, self.lstmfeat[{k,{}}]) > 0 then
               self.pebpxi:add(self.lstmfeat[{k,{}}]):add(- self.ins_lstmfeat)
               self.pebpvi:add(- self.ins_cnnfeat)
            end
            if Token2(margin, self.ins_lstmfeat, self.ins_cnnfeat, self.cnnfeat[{k,{}}]) > 0 then
               self.pebpxi:add(- self.ins_lstmfeat)
               self.pebpvi:add(self.cnnfeat[{k,{}}]):add(- self.ins_cnnfeat)
            end
         end
      end
   end

   if self.gpuid>=0 then
      assert(torch.type(self.pebpxi)=='torch.CudaTensor')
      assert(torch.type(self.pebpvi)=='torch.CudaTensor')
   end
   self.gradInput = { self.pebpxi, self.pebpvi }
   return self.gradInput
end
```
